mono/metadata/marshal.c

```diff
@@ -5934,54 +5847,41 @@ mono_marshal_asany (MonoObject *o, MonoMarshalNative string_encoding, int param_
 	case MONO_TYPE_U8:
 	case MONO_TYPE_R4:
 	case MONO_TYPE_R8:
-		return mono_object_unbox (o);
-		break;
+		return mono_handle_unbox_unsafe (o);
 	case MONO_TYPE_STRING:
 		switch (string_encoding) {
 		case MONO_NATIVE_LPWSTR:
-			return mono_marshal_string_to_utf16_copy ((MonoString*)o);
+			return mono_marshal_string_to_utf16_copy_handle (MONO_HANDLE_CAST (MonoString, o), error);
 		case MONO_NATIVE_LPSTR:
 		case MONO_NATIVE_UTF8STR:
 			// Same code path, because in Mono, we treated strings as Utf8
-			return mono_string_to_utf8str ((MonoString*)o);
+			return mono_string_handle_to_utf8 (MONO_HANDLE_CAST (MonoString, o), error);
```

Comment:
This is incorrect.
It should be `mono_string_to_utf8str_handle`.
This causes a  regression on Windows. https://github.com/mono/mono/issues/11898
Because it changes the allocator from CoTaskMemAlloc to malloc.
And then the later free CoTaskMemFree does not match.

The functions sound identical, as they are a combination of the same elements -- "string handle to utf8" but the different "spellings", underneath, are a different allocator.

It probably is ok on WIndows/amd64, because the underlying allocators are ultimately always HeapAlloc. It is whether the same heap is passed, i.e. GetProcessHeap or some other CreateHeap.

Reply:
This was only a problem in the debug build, because of the debug C runtime.
The retail C runtime's heap matches CoTaskMemAlloc.
What was really breaking in CI is  due to a gcc ABI bug.
See  https://github.com/mono/mono/pull/11917.